eras/shelley/impl/src/Cardano/Ledger/Shelley/Genesis.hs

```diff
@@ -129,6 +149,71 @@ emptyGenesisStaking =
       sgsStake = mempty
     }
 
+-- | Unlike @'NominalDiffTime'@ that supports @'Pico'@ precision, this type
+-- only supports @'Micro'@ precision.
+newtype NominalDiffTimeMicro = NominalDiffTimeMicro Micro
+  deriving (Show, Eq, Generic)
+  deriving anyclass (NoThunks)
+  deriving newtype (Ord, Num, Fractional, Real, ToJSON, FromJSON)
+
+deriving instance Generic Micro
+
+deriving anyclass instance NoThunks Micro
+
+instance ToCBOR NominalDiffTimeMicro where
+  toCBOR = encodeNominalDiffTimeMicro
+
+encodeNominalDiffTimeMicro :: NominalDiffTimeMicro -> Encoding
+encodeNominalDiffTimeMicro (NominalDiffTimeMicro micro) =
+  encodeInteger . toMicroseconds $ micro
+  where
+    toMicroseconds t =
+      numerator (toRational t * toRational (resolution $ Proxy @E6))
+
+instance FromCBOR NominalDiffTimeMicro where
+  fromCBOR = decodeNominalDiffTimeMicro
+
+decodeNominalDiffTimeMicro :: Decoder s NominalDiffTimeMicro
+decodeNominalDiffTimeMicro =
+  NominalDiffTimeMicro . fromRational . (% 1_000_000) <$> decodeInteger
+
+-- | There is no loss of resolution in this conversion
+microToPico :: Micro -> Pico
+microToPico micro = fromRational @Pico $ toRational micro
+
+-- | Loss of resolution occurs in this conversion
+picoToMicro :: Pico -> Micro
+picoToMicro pico = fromRational @Micro $ toRational pico
+
+fromNominalDiffTimeMicro :: NominalDiffTimeMicro -> NominalDiffTime
+fromNominalDiffTimeMicro =
+  secondsToNominalDiffTime . microToPico . nominalDiffTimeMicroToMicroseconds
+
+toNominalDiffTimeMicroWithRounding :: NominalDiffTime -> NominalDiffTimeMicro
+toNominalDiffTimeMicroWithRounding =
+  microsecondsToNominalDiffTimeMicro . picoToMicro . nominalDiffTimeToSeconds
+
+toNominalDiffTimeMicro :: NominalDiffTime -> Maybe NominalDiffTimeMicro
+toNominalDiffTimeMicro ndt
+  | roundedPicoseconds == picoseconds = Just ndtm
+  | otherwise = Nothing
+  where
+    ndtm@(NominalDiffTimeMicro microseconds) = toNominalDiffTimeMicroWithRounding ndt
+    picoseconds = nominalDiffTimeToSeconds ndt
+    roundedPicoseconds = microToPico microseconds
```

Comment:
This can also be simpler:
```suggestion
  | fromNominalDiffTimeMicro ndtm == ndt = Just ndtm
  | otherwise = Nothing
  where
    ndtm = toNominalDiffTimeMicroWithRounding ndt
```

Reply:
Awesome! Thanks! :bow: